pkg/ingester/ingester_v2.go

```diff
@@ -536,6 +537,88 @@ func createUserStats(db *userTSDB) *client.UserStatsResponse {
 	}
 }
 
+// v2QueryStream streams metrics from a TSDB. This implements the client.IngesterServer interface
+func (i *Ingester) v2QueryStream(req *client.QueryRequest, stream client.Ingester_QueryStreamServer) error {
+	_, ctx := spanlogger.New(stream.Context(), "v2QueryStream")
+
+	userID, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return err
+	}
+
+	from, through, matchers, err := client.FromQueryRequest(req)
+	if err != nil {
+		return err
+	}
+
+	i.metrics.queries.Inc()
+
+	db := i.getTSDB(userID)
+	if db == nil {
+		return nil
+	}
+
+	q, err := db.Querier(int64(from), int64(through))
+	if err != nil {
+		return err
+	}
+	defer q.Close()
+
+	ss, err := q.Select(matchers...)
+	if err != nil {
+		return err
+	}
+
+	timeseries := make([]client.TimeSeries, 0, queryStreamBatchSize)
+	batchSize := 0
+	numSamples := 0
+	numSeries := 0
+	for ss.Next() {
+		series := ss.At()
+
+		// convert labels to LabelAdapter
+		ts := client.TimeSeries{
+			Labels: client.FromLabelsToLabelAdapters(series.Labels()),
+		}
+
+		it := series.Iterator()
+		for it.Next() {
+			t, v := it.At()
+			ts.Samples = append(ts.Samples, client.Sample{Value: v, TimestampMs: t})
+		}
+		numSamples += len(ts.Samples)
+
+		timeseries = append(timeseries, ts)
+		numSeries++
+		batchSize++
+		if batchSize >= queryStreamBatchSize {
+			err = stream.Send(&client.QueryStreamResponse{
+				Timeseries: timeseries,
+			})
+			if err != nil {
+				return err
+			}
+
+			batchSize = 0
+			timeseries = timeseries[:0]
+		}
+	}
+
```

Comment:
At this point, we should check `ss.Err()` and return an error if `!= nil`, because the `for ss.Next()` could also stop iterating if an error is encountered.

Reply:
Added.